requirements.txt

```diff
@@ -1,20 +1,23 @@
-backcall==0.1.0
-decorator==4.4.0
-ipython==7.8.0
-ipython-genutils==0.2.0
-jedi==0.15.1
-numpy==1.17.3
-pandas==0.25.2
+attrs==19.3.0
```

Comment:
I think this has been explained somewhere but I'm still confused. Is requirements.txt the source of truth on this stuff? Put another way, is there any reason we also need `install_requires` inside setup.py?

Reply:
@johncmerfeld I was just reading answers on the [SO Post on the difference](https://stackoverflow.com/questions/43658870/requirements-txt-vs-setup-py/49684835) (and even asked a question myself in the comments). It looks like you can default to install_requires in your requirements.txt file but not the other way around. From the [official documentation](https://packaging.python.org/discussions/install-requires-vs-requirements/#requirements-files),

> Whereas install_requires requirements are minimal, requirements files often contain an exhaustive listing of pinned versions for the purpose of achieving repeatable installations of a complete environment.

I also found [this](https://packaging.python.org/discussions/install-requires-vs-requirements/): 

> Whereas install_requires metadata is automatically analyzed by pip during an install, requirements files are not, and only are used when a user specifically installs them using pip install -r .

I had originally planned on using pip in the CI/CD, but after seeing @kjmerf use requirements.txt, I looked around and saw the same thing on other people's pipelines, and I think it makes more sense to lock down the versions so if the code breaks all the sudden, we know it has to be a code change rather than a downstream package change.

We do need a plan I think to stop overwriting each other's requirements.txt! I should have probably installed from the existing one, but this one I created from a clean virtual environment. Let me know if this diatribe makes you feel better or worse.

Reply:
Better! Glad to know there's a meaningful difference and sounds like we're mostly concerned with `install_requires`. Thanks Ben!